src/reanimated2/helperTypes.ts

```diff
@@ -8,74 +7,58 @@ This will not be easy though!
 */
 
 import type {
-  ColorValue,
+  ImageStyle,
+  RegisteredStyle,
   StyleProp,
-  TransformsStyle as RNTransformsStyle,
+  TextStyle,
+  TransformsStyle,
+  ViewStyle,
 } from 'react-native';
+import type { AnimatableValue, SharedValue } from './commonTypes';
+import type { BaseAnimationBuilder } from './layoutReanimation/animationBuilder/BaseAnimationBuilder';
 import type {
-  AnimatableValue,
-  BaseAnimationBuilder,
   EntryExitAnimationFunction,
   LayoutAnimationFunction,
-  SharedValue,
-} from '.';
+} from './layoutReanimation/animationBuilder/commonTypes';
 import type { ReanimatedKeyframe } from './layoutReanimation/animationBuilder/Keyframe';
 import type { SharedTransition } from './layoutReanimation/sharedTransitions';
 import type { DependencyList } from './hook/commonTypes';
 
-export type Adaptable<T> =
-  | T
-  | ReadonlyArray<T | ReadonlyArray<T>>
-  | SharedValue<T>;
+export type TransformArrayItem = NonNullable<
+  TransformsStyle['transform']
+>[number];
 
-export type AdaptTransforms<T> = {
-  [P in keyof T]: Adaptable<T[P]>;
-};
-
-type TransformsStyle = Pick<RNTransformsStyle, 'transform'>;
+export type AnimatedTransform = MaybeSharedValueRecursive<
+  TransformsStyle['transform']
+>;
 
-export type TransformStyleTypes = TransformsStyle['transform'] extends
-  | readonly (infer T)[]
-  | string
-  | undefined
-  ? T
+type MaybeSharedValue<Value> = Value | Value extends AnimatableValue
+  ? SharedValue<Value>
   : never;
-export type AnimatedTransform = AdaptTransforms<TransformStyleTypes>[];
 
-/**
- * @deprecated Please use `AnimatedStyle` type instead.
- */
-export type AnimateStyle<S> = {
-  [K in keyof S]: K extends 'transform'
-    ? AnimatedTransform
-    : S[K] extends ReadonlyArray<any>
-    ? ReadonlyArray<AnimateStyle<S[K][0]>>
-    : S[K] extends object
-    ? AnimateStyle<S[K]>
-    : S[K] extends ColorValue | undefined
-    ? S[K] | number
-    : S[K] | SharedValue<AnimatableValue>;
-};
+type MaybeSharedValueRecursive<Value> = Value extends (infer InnerValue)[]
+  ?
+      | SharedValue<InnerValue[]>
+      | (MaybeSharedValueRecursive<InnerValue> | InnerValue)[]
+  : Value extends object
+  ?
+      | SharedValue<Value>
+      | { [K in keyof Value]: MaybeSharedValueRecursive<Value[K]> | Value[K] }
+  : MaybeSharedValue<Value>;
```

Comment:
Should `InnerValue[]` and `Value` be the same type?
So we could move the common part above the conditional.
WDYT about this type:
```
type MaybeSharedValueRecursive<Value> =
  | (Value extends AnimatableValue ? SharedValue<Value> : never)
  | (Value extends Array<infer ArrayElement>
      ? (MaybeSharedValueRecursive<ArrayElement> | ArrayElement)[]
      : Value extends object
      ? {
          [K in keyof Value]: MaybeSharedValueRecursive<Value[K]> | Value[K];
        }
      : Value);
      
```

I've noticed that according to definition of `MaybeSharedValue<Value>` we can reduce it to `Value` (since we already have `SharedValue<Value>` in our intersection.

Reply:
1. AnimatableValue definitely cannot be first because of this:
<img width="521" alt="Screenshot 2023-08-28 at 14 05 49" src="https://github.com/software-mansion/react-native-reanimated/assets/40713406/365ac712-bbae-4e21-a18e-5a915da5308d">

2. I was thinking about making this type more concise, but I think in it's current form it's much more transparent

Now we clearly see that you can **sharevaluelize** an array, array elements, an object, objects properties and some primitives.